File: src/algorithms/trees/bst-operations/bst-kth-smallest-iterative/sources/bst-kth-smallest-iterative_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.dirname(__file__))
module = importlib.import_module("bst-kth-smallest-iterative")
BSTNode = module.BSTNode
bst_kth_smallest_iterative = module.bst_kth_smallest_iterative


def make_node(value, left=None, right=None):
    node = BSTNode(value)
    node.left = left
    node.right = right
    return node


tree = make_node(4, make_node(2, make_node(1), make_node(3)), make_node(6, make_node(5), make_node(7)))


def test_first_smallest():
    assert bst_kth_smallest_iterative(tree, 1) == 1


def test_second_smallest():
    assert bst_kth_smallest_iterative(tree, 2) == 2


def test_seventh_smallest():
    assert bst_kth_smallest_iterative(tree, 7) == 7


def test_out_of_range():
    assert bst_kth_smallest_iterative(tree, 99) == -1


if __name__ == "__main__":
    test_first_smallest()
    test_second_smallest()
    test_seventh_smallest()
    test_out_of_range()
    print("All tests passed!")

File: src/algorithms/trees/bst-operations/bst-kth-smallest-iterative/__tests__/bst-kth-smallest-iterative_test.rs
include!("../sources/bst-kth-smallest-iterative.rs");

#[cfg(test)]
mod tests {
    use super::*;

    fn make_node(value: i32, left: Option<Box<BSTNode>>, right: Option<Box<BSTNode>>) -> Option<Box<BSTNode>> {
        Some(Box::new(BSTNode { value, left, right }))
    }

    fn leaf(value: i32) -> Option<Box<BSTNode>> {
        make_node(value, None, None)
    }

    fn build_tree() -> Option<Box<BSTNode>> {
        make_node(4,
            make_node(2, leaf(1), leaf(3)),
            make_node(6, leaf(5), leaf(7)),
        )
    }

    #[test]
    fn test_first_smallest() {
        assert_eq!(bst_kth_smallest_iterative(&build_tree(), 1), 1);
    }

    #[test]
    fn test_second_smallest() {
        assert_eq!(bst_kth_smallest_iterative(&build_tree(), 2), 2);
    }

    #[test]
    fn test_seventh_smallest() {
        assert_eq!(bst_kth_smallest_iterative(&build_tree(), 7), 7);
    }

    #[test]
    fn test_out_of_range() {
        assert_eq!(bst_kth_smallest_iterative(&build_tree(), 99), -1);
    }
}
